### ISSUE TITLE
False negatives in 3 rules when right side of && expression is a literal

### ISSUE THREAD

**Issue by btmills** [eslint/eslint#13634]
**Tell us about your environment**

* **ESLint Version:** 7.7.0
* **Node Version:** 14.9.0
* **npm Version:** 6.14.7

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    parserOptions: {
        ecmaVersion: 2020
    }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
/* eslint constructor-super: "error",
   no-throw-literal: "error",
   prefer-promise-reject-errors: "error"
 */
let anything;

// If anything is a constructor, it'll be truthy, we'll evaluate the
// right side of the &&, and we'll extend 42.
class Subclass extends (anything && 42) {
    constructor() {
        super();
    }
}

// If anything is an Error instance, it'll be truthy, we'll evaluate
// the right side of the &&, and we'll throw 42.
throw anything && 42;

// If anything is an Error instance, it'll be truthy, we'll evaluate
// the right side of the &&, and we'll reject with 42.
Promise.reject(anything && 5);
```

[Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IGNvbnN0cnVjdG9yLXN1cGVyOiBcImVycm9yXCIsXG4gICBuby10aHJvdy1saXRlcmFsOiBcImVycm9yXCIsXG4gICBwcmVmZXItcHJvbWlzZS1yZWplY3QtZXJyb3JzOiBcImVycm9yXCJcbiAqL1xubGV0IGFueXRoaW5nO1xuXG4vLyBJZiBhbnl0aGluZyBpcyBhIGNvbnN0cnVjdG9yLCBpdCdsbCBiZSB0cnV0aHksIHdlJ2xsIGV2YWx1YXRlIHRoZVxuLy8gcmlnaHQgc2lkZSBvZiB0aGUgJiYsIGFuZCB3ZSdsbCBleHRlbmQgNDIuXG5jbGFzcyBTdWJjbGFzcyBleHRlbmRzIChhbnl0aGluZyAmJiA0Mikge1xuICAgIGNvbnN0cnVjdG9yKCkge1xuICAgICAgICBzdXBlcigpO1xuICAgIH1cbn1cblxuLy8gSWYgYW55dGhpbmcgaXMgYW4gRXJyb3IgaW5zdGFuY2UsIGl0J2xsIGJlIHRydXRoeSwgd2UnbGwgZXZhbHVhdGVcbi8vIHRoZSByaWdodCBzaWRlIG9mIHRoZSAmJiwgYW5kIHdlJ2xsIHRocm93IDQyLlxudGhyb3cgYW55dGhpbmcgJiYgNDI7XG5cbi8vIElmIGFueXRoaW5nIGlzIGFuIEVycm9yIGluc3RhbmNlLCBpdCdsbCBiZSB0cnV0aHksIHdlJ2xsIGV2YWx1YXRlXG4vLyB0aGUgcmlnaHQgc2lkZSBvZiB0aGUgJiYsIGFuZCB3ZSdsbCByZWplY3Qgd2l0aCA0Mi5cblByb21pc2UucmVqZWN0KGFueXRoaW5nICYmIDUpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**What did you expect to happen?**

We should report errors in each of the above cases. All three are looking for a specific type of truthy value from the expression. Because the expression uses `&&`, any possibly-valid value would evaluate to the right side of the expression, which would not be a valid value in these examples.

This was originally discovered in https://github.com/eslint/eslint/pull/13618#discussion_r479584827. Because it relates to existing behavior that existed prior to that pull request, we decided to split it into its own issue that can be fixed after the pull request is merged. This bug is a false negative in the 3 rules above, and fixing it would cause them to report more errors. Does anyone object to fixing this in a semver-minor change as allowed by our semver policy?

The fix would update both [`astUtils.couldBeError()`](https://github.com/eslint/eslint/blob/a32032430a0779a4e3b2d137d4d0682844084b82/lib/rules/utils/ast-utils.js#L1587-L1588) and [`constructor-super`'s `isPossibleConstructor()`](https://github.com/eslint/eslint/blob/a32032430a0779a4e3b2d137d4d0682844084b82/lib/rules/constructor-super.js#L65-L69) to only check the right side of `&&` `LogicalExpression`s.

**What actually happened? Please include the actual, raw output from ESLint.**

No errors are reported.

**Are you willing to submit a pull request to fix this bug?**

Yes, if @mdjermanovic doesn't get to it first.

**Comment by mdjermanovic** [comment]
I support fixing this in a semver-minor change.
